### PR TITLE
Unit test fixes

### PR DESCRIPTION
**Pull request by grahamm-hexagon**
WIP changes for unit-test

### CODE REVIEW

**Comment by yaseen-mowzer-hexagon** [comment]
I just want to double check that we only made changes to the tests. There were no changes to the actual ProcessFamily necessary?

**Comment by yaseen-mowzer-hexagon** [comment]
Can you add a summary of your analysis (e.g. which tests are failing and why that is okay) in the pull request description?

File: processfamily/test_processfamily.py [hunk 1]
```diff
@@ -24,86 +27,249 @@
...
-    def process_exists_or_access_denied(self, pid):
+
+def assert_middle_child_port_unbound():
+    port = Config.get_starting_port_nr() + 2
+    logging.info("Checking for ability to bind to port %d", port)
+    serversocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        #if not sys.platform.startswith('win'):
+            # On linux I need this setting cos we are starting and stopping things
+            # so frequently that they are still in a STOP_WAIT state when I get here
```
Comment (yaseen-mowzer-hexagon) [inline processfamily/test_processfamily.py:69]: Can you delete the part of the comment that reference windows and linux. I think that just makes it confusing.

**Review by yaseen-mowzer-hexagon — CHANGES_REQUESTED** [review]

**Comment by grahamm-hexagon** [comment]
> 

correct

**Comment by grahamm-hexagon** [comment]
Adding this as a comment.
The following tests are failing.
When running the testcases individually the Windows service tests are failing when testing with a different user.  This appears to be environmental.
When testing some of the startup edge cases, the tests do not count the number of terminated processes correctly.

**Review by yaseen-mowzer-hexagon — APPROVED** [review]

**Comment by yaseen-mowzer-hexagon** [comment]
From our internal JIRA ticket, this is what tests are failing:
```
test_processfamily.py::TestFunkyWebServer::test_parent_stop_child_locked_up[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_parent_exit_child_locked_up[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_parent_kill_child_locked_up[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_exit_on_start[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_error_during_run[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_on_start[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_error_on_start[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_error_during_init[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_during_init[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_crash_on_start[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_file_open_by_parent_before_fork_can_be_closed_and_deleted[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_echo_std_err_on[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_handles_over_commandline_off[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_handles_over_commandline_off_close_fds_off[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_close_fds_off[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_comms_strategy_stdin_close[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_comms_strategy_none[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_child_comms_strategy_signal[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_use_job_object_off[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_cpu_affinity_off[windows_service-None]
test_processfamily.py::TestFunkyWebServer::test_handles_over_commandline_off_file_open_by_parent[windows_service-None]
test_processfamily.py::TestWindowsService::test_service_stop[None]
test_processfamily.py::TestWindowsService::test_service_stop_child_locked_up[None]
test_processfamily.py::TestWindowsService::test_service_stop_child_freeze_on_start[None]
test_processfamily.py::TestFunkyWebServer::test_parent_stop[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_parent_exit[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_parent_stop_child_locked_up[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_parent_exit_child_locked_up[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_exit_on_start[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_error_during_run[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_on_start[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_error_on_start[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_error_during_init[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_during_init[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_crash_on_start[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_file_open_by_parent_before_fork_can_be_closed_and_deleted[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_echo_std_err_on[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_handles_over_commandline_off[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_handles_over_commandline_off_close_fds_off[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_close_fds_off[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_comms_strategy_stdin_close[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_comms_strategy_none[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_comms_strategy_signal[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_use_job_object_off[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_cpu_affinity_off[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_handles_over_commandline_off_file_open_by_parent[windows_service-NT AUTHORITY\NetworkService]
test_processfamily.py::TestWindowsService::test_service_stop[NT AUTHORITY\NetworkService]
test_processfamily.py::TestWindowsService::test_service_stop_child_locked_up[NT AUTHORITY\NetworkService]
test_processfamily.py::TestWindowsService::test_service_stop_child_freeze_on_start[NT AUTHORITY\NetworkService]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_on_start[normal_subprocess]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_on_start[pythonw]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_during_init[normal_subprocess]
test_processfamily.py::TestFunkyWebServer::test_child_freeze_during_init[pythonw]
```
